bundles/org.openhab.core/src/test/java/org/openhab/core/library/items/NumberItemTest.java

```diff
@@ -209,4 +210,61 @@ public void testKelvinToMired() {
 
         assertThat(item.getState().format("%.0f mired"), is("370 mired"));
     }
+
+    @Test
+    public void stateDescriptionUnitUsedWhenStateDescriptionPresent() {
+        UnitProvider unitProviderMock = mock(UnitProvider.class);
+        when(unitProviderMock.getUnit(Temperature.class)).thenReturn(SIUnits.CELSIUS);
+        when(stateDescriptionServiceMock.getStateDescription(ITEM_NAME, null)).thenReturn(
+                StateDescriptionFragmentBuilder.create().withPattern("%.0f °F").build().toStateDescription());
+
+        NumberItem item = new NumberItem("Number:Temperature", ITEM_NAME);
+        item.setStateDescriptionService(stateDescriptionServiceMock);
+        item.setUnitProvider(unitProviderMock);
+
+        assertThat(item.getUnit(), is(ImperialUnits.FAHRENHEIT));
+
+        item.setState(new QuantityType<>("429 °F"));
+        assertThat(item.getState(), is(new QuantityType<>("429 °F")));
+
+        item.setState(new QuantityType<>("165 °C"));
+        assertThat(item.getState(), is(new QuantityType<>("329 °F")));
+    }
+
+    @Test
+    public void unitPreservedWhenStateDescriptionContainsWildCard() {
+        UnitProvider unitProviderMock = mock(UnitProvider.class);
+        when(unitProviderMock.getUnit(Temperature.class)).thenReturn(SIUnits.CELSIUS);
+        when(stateDescriptionServiceMock.getStateDescription(ITEM_NAME, null)).thenReturn(
+                StateDescriptionFragmentBuilder.create().withPattern("%.0f %unit%").build().toStateDescription());
+
+        NumberItem item = new NumberItem("Number:Temperature", ITEM_NAME);
+        item.setStateDescriptionService(stateDescriptionServiceMock);
+        item.setUnitProvider(unitProviderMock);
+
+        assertThat(item.getUnit(), is(nullValue()));
+
+        item.setState(new QuantityType<>("329 °F"));
+        assertThat(item.getState(), is(new QuantityType<>("329 °F")));
+
+        item.setState(new QuantityType<>("100 °C"));
+        assertThat(item.getState(), is(new QuantityType<>("100 °C")));
+    }
+
+    @Test
+    public void defaultUnitUsedWhenStateDescriptionEmpty() {
+        UnitProvider unitProviderMock = mock(UnitProvider.class);
+        when(unitProviderMock.getUnit(Temperature.class)).thenReturn(SIUnits.CELSIUS);
+
+        NumberItem item = new NumberItem("Number:Temperature", ITEM_NAME);
+        item.setUnitProvider(unitProviderMock);
+
+        assertThat(item.getUnit(), is(SIUnits.CELSIUS));
+
+        item.setState(new QuantityType<>("329 °F"));
+        assertThat(item.getState(), is(new QuantityType<>("165 °C")));
+
+        item.setState(new QuantityType<>("100 °C"));
+        assertThat(item.getState(), is(new QuantityType<>("100 °C")));
+    }
```

Comment:
I've added an additional unit test to test specifically for not converting kW to Joule when placeholder is used.
```suggestion
    }
    
    @Test
    void testNoUnitWhenUnitPlaceholderUsed() {
        final UnitProvider unitProviderMock = mock(UnitProvider.class);
        when(unitProviderMock.getUnit(Energy.class)).thenReturn(Units.JOULE);

        final NumberItem item = new NumberItem("Number:Energy", ITEM_NAME);
        item.setUnitProvider(unitProviderMock);
        assertThat(item.getUnit(), is(Units.JOULE));

        item.setStateDescriptionService(stateDescriptionServiceMock);
        item.setState(new QuantityType<>("329 kWh"));
        assertThat(item.getState(), is(new QuantityType<>("329 kWh")));
        assertNull(item.getUnit());
    }
```